tstest/integration/vms/vms_test.go

```diff
@@ -0,0 +1,457 @@
+		{"opensuse-leap-15.2", "10", "https://download.opensuse.org/repositories/Cloud:/Images:/Leap_15.2/images/openSUSE-Leap-15.2-OpenStack.x86_64.qcow2", "4df9cee9281d1f57d20f79dc65d76e255592b904760e73c0dd44ac753a54330f", 512},
+		{"opensuse-tumbleweed", "5", "https://download.opensuse.org/tumbleweed/appliances/openSUSE-Tumbleweed-JeOS.x86_64-OpenStack-Cloud.qcow2", "f5a40c693187dddc524b0877b0aba31cdb6f2e78f12acce0bbb2975ec120dfc0", 512},
+		{"ubuntu-16.04", "5", "https://cloud-images.ubuntu.com/xenial/current/xenial-server-cloudimg-amd64-disk1.img", "50a21bc067c05e0c73bf5d8727ab61152340d93073b3dc32eff18b626f7d813b", 512},
+		{"ubuntu-18.04", "5", "https://cloud-images.ubuntu.com/bionic/current/bionic-server-cloudimg-amd64.img", "78f09b6086367914d23f26e38c3cd88a2aaf1a6a4500ebd46c74ad60c56c1585", 512},
+		{"ubuntu-20.04", "5", "https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64.img", "513158b22ff0f08d0a078d8d60293bcddffdb17094a7809c76c52aba415ecc54", 512},
+		{"ubuntu-20.10", "5", "https://cloud-images.ubuntu.com/groovy/current/groovy-server-cloudimg-amd64.img", "c1332c24557389a129ff98fa169e34cb53c02555ed702a235e26b8978dd004c3", 512},
+		{"ubuntu-21.04", "5", "https://cloud-images.ubuntu.com/hirsute/current/hirsute-server-cloudimg-amd64.img", "2f8a562637340a026f712594f1257673543d74725d8e3daf88d533d7b8bf448f", 512},
+	}
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatalf("can't getwd: %v", err)
+	}
+
+	dir, err := os.MkdirTemp(cwd, "vmIntegrationTest")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(dir)
+
+	lis, err := net.Listen("tcp", "192.168.122.1:0")
+	if err != nil {
+		t.Fatalf("can't make TCP listener: %v", err)
+	}
+	defer lis.Close()
+	t.Logf("host:port: %s", lis.Addr())
+
+	cs := &testcontrol.Server{}
+
+	var ipLock sync.Mutex
+	ipMap := map[string]string{}
+
+	mux := http.NewServeMux()
+	mux.Handle("/", cs)
+	mux.HandleFunc("/myip/", func(w http.ResponseWriter, r *http.Request) {
+		ipLock.Lock()
+		defer ipLock.Unlock()
+
+		name := path.Base(r.URL.Path)
+		host, _, _ := net.SplitHostPort(r.RemoteAddr)
+		ipMap[name] = host
+		t.Logf("%s: %v", name, host)
+	})
+
+	hs := &http.Server{Handler: mux}
+	go hs.Serve(lis)
+
+	run(t, dir, "ssh-keygen", "-t", "ed25519", "-f", "machinekey", "-N", ``)
+	pubkey, err := os.ReadFile(filepath.Join(dir, "machinekey.pub"))
+	if err != nil {
+		t.Fatalf("can't read ssh key: %v", err)
+	}
+
+	loginServer := fmt.Sprintf("http://%s", lis.Addr())
+
+	cancels := make(chan func(), len(distros))
+
+	t.Run("mkvm", func(t *testing.T) {
+		for n, distro := range distros {
+			n, distro := n, distro
+			t.Run(distro.name, func(t *testing.T) {
+				t.Parallel()
+
+				cancel := mkVM(t, n, distro, string(pubkey), loginServer, dir)
+				cancels <- cancel
+			})
+		}
+	})
+
+	close(cancels)
+	for cancel := range cancels {
+		//lint:ignore SA9001 They do actually get ran
+		defer cancel()
+
+		t.Logf("%d vms in the pipe", len(cancels))
+
+		if len(cancels) == 0 {
```

Comment:
how can this ever be true? it's in a range over cancels, so it has to be non-zero here?

Reply:
oh! it's a chan you're ranging over.

don't use len on a chan concurrently with sending to it. that's potentialy racey.

Reply:
Based on the execution flow it should be impossible for this to be racy. The buffered channel will be completely written to by the point that this code runs. I'm using a channel as a deque here.

Reply:
Go convention for this is to instead do a `for i := 0; i < len(distros); i++ { // read from chan etc }` when you know exactly how many elements you expect, as is the case here. I think you're right that using len here is safe, but doing it the conventional way will avoid go veteran spidey senses going off.